### PR TITLE
Add Bubble Sort in Typescript

### LINKED ISSUE #3366: Add Bubble Sort in Typescript
To request a new code snippet, please fill out the following:

Project name: Bubble Sort
Project link: https://sampleprograms.io/projects/bubble-sort
Language: Typescript

> The project link must be a link to a project on the [Sample Programs Website project list][sample-programs-project-list].
> If you would like to add a new project to the Sample Programs Website,
> please first make a pull request to the [Sample Programs Website][sample-programs-website].

[sample-programs-website]: https://github.com/TheRenegadeCoder/sample-programs-website
[sample-programs-project-list]: https://sampleprograms.io/projects/

### PR DESCRIPTION
**Pull request by heishmanA**
Congrats on taking the first step to contributing to the Sample Programs repository maintained by [The Renegade Coder][renegade-coder]! 
For simplicity, please make sure that your pull request includes one and only one contribution.

Please fill _one_ of the sections below as applicable.
Please also add any other relevant information to the Notes section at the bottom.
You may delete or just ignore any other sections.
For more information please refer to our [contributing documentation][contributing]

## I Am Adding a New Code Snippet in an Existing Language

- [x] I fixed #3366 
- [x] I named the pull request using `Add {PROJECT} in {LANGUAGE}` format
  
  
## I Am Adding a New Code Snippet in a New Language

- [ ] I fixed #your-issue-number-here
- [ ] I named the pull request using `Add {PROJECT} in {LANGUAGE}` format
- [ ] I added a `testinfo.yml` files (see [contributing documentation][contributing-new-language])
  - [ ] I used an officially supported docker image or one that I personally trust
  

## I Am Adding New Tests for a Project

- [ ] I fixed #your-issue-number-here
- [ ] I named the pull request using `Add {PROJECT} Tests` format
- [ ] I verified the project exists on the [Sample Programs Project List][sample-programs-project-list]
- [ ] I added tests for every test case in the Testing table of the project description for the given project. (See [contributing documentation][contributing-tests-in-detail])

## I Am Adding New Tests for a Language

- [ ] I fixed #your-issue-number-here
- [ ] I named the pull request using `Add {LANGUAGE} Tests` format
- [ ] I added a `testinfo.yml` files (see [contributing documentation][contributing-new-language])
  - [ ] I used an officially supported docker image or one that I personally trust
- [ ] I verified all tests are passing

  
## I Am Modifying an Existing Code Snippet or Existing Tests

- [ ] I fixed #your-issue-number-here
- [ ] I have read about modifying existing code snippets or tests in the [contributing documentation][contributing-modifications]
- [ ] I have explained why this modification is _necessary_ either in the issue linked above or in the notes below
- [ ] I named the pull request using `Modify {PROJECT} {tests} in {LANGUAGE}` format

## Other Notes

Add notes here...

[renegade-coder]: https://therenegadecoder.com/
[contributing-plagiarism]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#plagiarism
[contributing-new-project]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#requirements-for-a-new-project
[contributing-new-language]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#requirements-for-a-new-language
[contributing-readme]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#create-readmes
[contributing-tests-in-detail]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#tests-in-detail
[contributing]: ../CONTRIBUTING.md
[sample-programs-project-list]: https://sampleprograms.io/projects/
[contributing-modifications]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#modifying-existing-code-snippets

### CODE REVIEW

File: archive/t/typescript/bubble-sort.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    const end = array.length;
+    var i = 0,
+        j = 1,
+        start = 0,
+        done = 0;
+    while (done < end) {
+        /* Swap returns 0 to signify that bubble sort should do another pass */
+        done = Compare(array, i, j) ? Swap(array, i, j) : done + 1;
+        /* left value index - cycles between 0 and length - 1 of the array. */
+        i = i < end - 1 ? (i + 1) % end : start;
```
Comment (rzuckerm) [inline archive/t/typescript/bubble-sort.ts:67]: Seem like this would do the same this:
```suggestion
        i = (i + 1) % end;
```

**Review by rzuckerm — COMMENTED** [review]
I'm not convinced that this actually is a faithful implementation of the bubble sort. It appears to be similar to [this](https://en.wikipedia.org/wiki/Bubble_sort#Optimizing_bubble_sort), but I'm not entirely sure.

File: archive/t/typescript/bubble-sort.ts (hunk 1 above)
Comment (heishmanA) [reply archive/t/typescript/bubble-sort.ts:67]: You are absolutely correct. I was too worried about an index error from j = i + 1 and forgot that that shouldn't happen.

**Comment by heishmanA** [comment]
> I'm not convinced that this actually is a faithful implementation of the bubble sort. It appears to be similar to [this](https://en.wikipedia.org/wiki/Bubble_sort#Optimizing_bubble_sort), but I'm not entirely sure.

Oh, did you want the actual algorithm? I was just trying to avoid using a nested list.

**Comment by rzuckerm** [comment]
> Oh, did you want the actual algorithm? I was just trying to avoid using a nested loop.
> 
> -- Edit --
> 
> After looking over the algorithm and comparing I would have to say you're correct and my implementation isn't true to Bubble Sort. I'll rewrite it and place another pull request and follow the alogirthm this time!

Yes, this intended to be an actual implementation of the Bubble Sort Algorithm. You don't have to submit a new PR. You can just modify the code in `archive/t/typescript/bubble-sort.ts`.

**Comment by heishmanA** [comment]
> > Oh, did you want the actual algorithm? I was just trying to avoid using a nested loop.
> > -- Edit --
> > After looking over the algorithm and comparing I would have to say you're correct and my implementation isn't true to Bubble Sort. I'll rewrite it and place another pull request and follow the alogirthm this time!
> 
> Yes, this intended to be an actual implementation of the Bubble Sort Algorithm. You don't have to submit a new PR. You can just modify the code in `archive/t/typescript/bubble-sort.ts` for this PR.

Done!

**Review by rzuckerm — APPROVED** [review]
Thanks for making the requested changes. Looks good!

**Pull request merged**